docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,214 @@
+
+- Entirely eliminate direct usage of any Azure Go SDK
+- Reconcile long-running Azure operations asynchronously by waiting for
+  Kubernetes watch events from changes to the `status` of ASO resources instead
+  of periodically checking if the operation is complete
+- Reconcile different Azure resource types for the same workload cluster in parallel
+- Allow users to interact with CAPZ-created ASO resources directly, out-of-band with CAPZ
+- Adopt Azure Workload Identity to authenticate with Azure APIs
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - Users of new Azure features
+
+As a CAPZ user I want to take advantage of the latest and greatest features in Azure only available in the newest Azure API versions in my clusters managed by CAPZ. This would allow me to adopt CAPZ for a wider range of use cases and continue evolving my CAPZ clusters alongside the Azure API.
+
+#### Story 2 - Developer experience
+
+As a CAPZ developer I would like to only need to interact with the Kubernetes API server using Custom Resource Definitions representing Azure resources rather than having to deal with the impedance of interacting with the Azure APIs. This would enable me to focus more on the core value of CAPZ, transforming Cluster API resource definitions to Azure resource definitions.
+
+### Implementation Details/Notes/Constraints
+
+#### Reconciliation
+
+To minimize impact initially, as much of the existing pattern in `/azure/services` for interfacing with Azure services will be reused. The biggest change will involve a new wrapper adjacent to `/azure/services/async` such as `/azure/services/aso` which will handle functionality common to all ASO resources. The following diagram is based on the [async reconciliation diagram](./images/async-reconcile.png) and shows how resources will be reconciled with ASO:
+
+![Figure 1](./images/aso-reconcile.png)
+
+Similarly for the [async delete diagram](./images/async-delete.png), the following diagram shows how resources will be deleted with ASO:
+
+![Figure 2](./images/aso-delete.png)
+
+A proof-of-concept of the reconciliation flow has been implemented for Azure resource groups and is available here: https://github.com/nojnhuh/cluster-api-provider-azure/tree/aso/azure/services/asogroups
+
+##### BYO Resources
+
+BYO resources likely will not exist as ASO resources, so determining whether or not a given expected Azure resource exists becomes less straightforward using ASO. Initially, both BYO and CAPZ-managed resources will be created in ASO with ASO's `serviceoperator.azure.com/reconcile-policy: skip` label which tells ASO not to make any changes to the corresponding Azure resource, essentially assuming the resource is BYO. Then, if a future CAPZ reconciliation reveals the Azure resource does not exist, CAPZ will change it's assumption and make ASO "adopt" the resource by setting `serviceoperator.azure.com/reconcile-policy: manage` to allow changes made to the ASO resource to be reflected in Azure.
```

Comment:
> Then, if a future CAPZ reconciliation reveals the Azure resource does not exist

Can you expand on how this part works? Does CAPZ do an Azure API GET on that resource?

Reply:
Yeah I'll go into more details here. tl;dr is ASO does the GET and will populate the `status` with an `AzureResourceNotFound` condition that CAPZ consumes.